File: Suyeon/ch14_Tree/104_Maximum_Depth_of_Binary_tree.py
import collections

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:

        def maxheight(root):
            if root is None:
                return 0
            return max(maxheight(root.left), maxheight(root.right)) + 1


class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        queue = collections.deque()

        if root is None: #루트예외처리
            return 0
        queue.append(root)
        depth = 0
        while queue:
            depth += 1
            for _ in range(len(queue)): # 부모노드만큼만 돌기위해 처음엔 queue엔 항상 부모만 들어있다 이후에 섞인다
                p = queue.popleft() # p가 none인것이 애초에 들어올수없게
                if p.left is not None: # none이 아닌 자식들만 추가
                    queue.append(p.left)
                if p.right is not None:
                    queue.append(p.right)

        return depth

File: Sujin/Ch13_Graph&Tree&Dijkstra/104_Maximum_Depth_of_Binary_Tree.cpp
#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int res = 0;
    void DFS(TreeNode* n, int sum){
        if(sum > res) res = sum; // res에 sum의 최댓값을 계속 갱신하기
        
        if(n->left!=NULL) DFS(n->left, sum+1);
        if(n->right!=NULL) DFS(n->right, sum+1);
    }
    
    int maxDepth(TreeNode* root) {
        if(root==NULL) return 0; // 예외처리
        
        DFS(root, 1);
        return res;
    }
};